pkg/consensus/recovery_message.go

```diff
@@ -0,0 +1,172 @@
+package consensus
+
+import (
+	"github.com/CityOfZion/neo-go/pkg/io"
+	"github.com/CityOfZion/neo-go/pkg/util"
+	"github.com/pkg/errors"
+)
+
+type (
+	recoveryMessage struct {
+		PreparationHash     *util.Uint256
+		PreparationPayloads map[uint16]*preparationCompact
+		CommitPayloads      map[uint16]*commitCompact
+		ChangeViewPayloads  map[uint16]*changeViewCompact
+		PrepareRequest      *prepareRequest
+	}
+
+	changeViewCompact struct {
+		ValidatorIndex     uint16
+		OriginalViewNumber byte
+		Timestamp          uint32
+		InvocationScript   []byte
+	}
+
+	commitCompact struct {
+		ViewNumber       byte
+		ValidatorIndex   uint16
+		Signature        []byte
+		InvocationScript []byte
+	}
+
+	preparationCompact struct {
+		ValidatorIndex   uint16
+		InvocationScript []byte
+	}
+)
+
+const uint256size = 32
+
+func (m *recoveryMessage) DecodeBinary(r *io.BinReader) {
+	l := r.ReadVarUint()
+	m.ChangeViewPayloads = make(map[uint16]*changeViewCompact, l)
+	for i := uint64(0); i < l; i++ {
+		var p changeViewCompact
+		p.DecodeBinary(r)
+		m.ChangeViewPayloads[p.ValidatorIndex] = &p
+	}
+
+	var hasReq bool
+	r.ReadLE(&hasReq)
+	if hasReq {
+		m.PrepareRequest = new(prepareRequest)
+		m.PrepareRequest.DecodeBinary(r)
+	} else {
+		l := r.ReadVarUint()
+		if l != 0 {
+			if l == uint256size {
+				m.PreparationHash = new(util.Uint256)
+				r.ReadBE(m.PreparationHash[:])
+			} else {
+				r.Err = errors.New("invalid data")
+			}
+		} else {
+			m.PreparationHash = nil
+		}
+	}
+
+	l = r.ReadVarUint()
+	m.PreparationPayloads = make(map[uint16]*preparationCompact, l)
+	for i := uint64(0); i < l; i++ {
+		var p preparationCompact
+		p.DecodeBinary(r)
+		m.PreparationPayloads[p.ValidatorIndex] = &p
+	}
+
+	l = r.ReadVarUint()
+	m.CommitPayloads = make(map[uint16]*commitCompact, l)
+	for i := uint64(0); i < l; i++ {
+		var p commitCompact
+		p.DecodeBinary(r)
+		m.CommitPayloads[p.ValidatorIndex] = &p
+	}
+}
+
+func (m *recoveryMessage) EncodeBinary(w *io.BinWriter) {
+	w.WriteVarUint(uint64(len(m.ChangeViewPayloads)))
+	for _, p := range m.ChangeViewPayloads {
+		p.EncodeBinary(w)
+	}
+
+	hasReq := m.PrepareRequest != nil
+	w.WriteLE(hasReq)
+	if hasReq {
+		m.PrepareRequest.EncodeBinary(w)
+	} else {
+		if m.PreparationHash == nil {
+			w.WriteVarUint(0)
+		} else {
+			w.WriteVarUint(uint256size)
+			w.WriteBE(m.PreparationHash[:])
+		}
+	}
+
+	w.WriteVarUint(uint64(len(m.PreparationPayloads)))
```

Comment:
lines 104-107 and 109-112 look very similar. What about moving to some reusable function?

Reply:
even 86-89 same

Reply:
We have quite a lot of code like this, maybe it can be generalized in `io` as
```
func (r *io.Reader) ReadArrayOfSerializable(aos []io.Serializable)
```

Reply:
And BTW, we have `GetVarSize()` properly counting things like that.